Redirect non-admin users to core's root_path

andreslucena opened the pull request:
<!--
NOTE: We are in the middle of a big redesign of the frontend.
That could mean that your PR will not get through on the next few months.
Please see https://github.com/decidim/decidim/discussions/9512
-->

#### :tophat: What? Why?

While reviewing #10117, @ahukkanen [found out this issue](https://github.com/decidim/decidim/pull/10117#discussion_r1219019026):

> We need one more condition here because now regular users would be redirected to a 404 page in case they try to access an admin view.

I preferred to handle this in another PR just in case there are lots of specs to change.

#### :pushpin: Related Issues
 
- Related to #10117

#### Testing

~1. Sign in as Assembly's Admin ("assembly_1_admin@example.org")~
~2. Go to an admin page that you're not authorized, such as http://localhost:3000/admin/participatory_processes~
~3. See that you're redirected to the core's root_path ("/") instead of the admin's root_path ("/admin")~
(This was not actually a bug, [read my comment with the full explanation](https://github.com/decidim/decidim/pull/10968#issuecomment-1647408497))

Without the patch:

1. Login as `user@example.org`
2. Type in the URL `/admin/templates/questionnaire_templates`
3. Get redirected to the admin TOS page which gives you a 404
4. Apply the patch
5. See that you're redirected to the /admin page

:hearts: Thank you!

andreslucena commented on decidim-admin/app/controllers/decidim/admin/application_controller.rb:
```diff
@@ -42,6 +42,8 @@ class ApplicationController < ::DecidimController
                            ::Decidim::Admin::Permissions)
 
       def user_has_no_permission_path
+        return decidim.root_path if user_signed_in? && !current_user.admin?
```
Yes, I'll add it of course! This was just the initial change to see if I actually needed to make lots of changes in specs

andreslucena replied:
> This was just the initial change to see if I actually needed to make lots of changes in specs

Finally I didn't need to do any change to the specs. My guess is that all the specs were checking for "You are not authorized to perform this action" string and not for actually which path they were redirected to. 

Regarding the spec, it is available in 0db6e5b309b48ec53a966b1acd9b8b0df2460183

andreslucena commented:
While working on this spec I found another issue. As it's out of the scope and I don't have time to work on it now, I left it reported at #11004

andreslucena commented:
This is ready to be reviewed again @alecslupu

alecslupu commented:
I have tried this PR using latest branch, and i still got the 404 ... 

Tested with user@example.... and url:  http://localhost:3000/admin/assemblies/polish-thank/components

Last query before error: 
```
17:59:47 web.1       |   CACHE Decidim::Initiative Load (0.0ms)  SELECT "decidim_initiatives".* FROM "decidim_initiatives" INNER JOIN "decidim_initiatives_committee_members" ON "decidim_initiatives_committee_members"."decidim_initiatives_id" = "decidim_initiatives"."id" WHERE (decidim_initiatives_committee_members.state = 2) AND "decidim_initiatives_committee_members"."decidim_users_id" = $1 
```

andreslucena commented:
After reviewing again all the comments related to this issue and giving it a second thought with new eyes (as several weeks have passed), I think that we're talking about different issues and behaviors, mostly related but not really 😄 

## The original report

Comes from https://github.com/decidim/decidim/pull/10117#discussion_r1219019026

1. Login as `user@example.org`
2. Type in the URL `/admin/templates/questionnaire_templates`
3. Get redirected to the admin TOS page which gives you a 404

This actually was still happening. It's fixed with a77e6240bb92557e3b29d5b9e494b04ff783acc8

## My original testing explanation

Comes from https://github.com/decidim/decidim/pull/10968#issue-1747066823 

1. Sign in as Assembly's Admin ("assembly_1_admin@example.org")
2. Go to an admin page that you're not authorized, such as http://localhost:3000/admin/participatory_processes
3. See that you're redirected to the core's root_path ("/") instead of the admin's root_path ("/admin")

This is actually not a bug, as it's an admin-kind of user: it's an assembly's admin, it should have access to /admin, so it's ok that's redirected to /admin with the "Not authorized" callout.

I've changed it and also changed the specs that I've added to better reflect this behavior at 04a2089c867d7a6da7cd733323a0d510b66e5502 

## The 404 of a regular user in the admin panel

Comes from https://github.com/decidim/decidim/pull/10968#issuecomment-1634411706

1. Login as `user@example.org`
2. Go to an admin page that you're not authorized, such as http://localhost:3000/admin/assemblies/polish-thank/components
3. See that you have a 404 page

This is actually not a bug. If we showed other kind of errors here (like 403 Forbidden or 401 Unauthorized) we could be leaking some information about which spaces we have in the admin panel. For instance, http://localhost:3000/admin/assemblies/a-very-private-and-secret-process/components

----

So, while now I've fixed the original bug report (aka "The original report"), what I found is that **decidim-templates isn't consistent with the rest of modules when checking permissions/routes**. Basically this is visible by following the next steps:

1. Login as `user@example.org`
2. Type in the URL `/admin/assemblies/exaggerate-plain/edit` 
3. See that you have a 404 page with this same URL
4. Type in the URL  `/admin/templates/questionnaire_templates`
5. See that you have a 404 page - but now we're redirected to the /admin path.

As this was not detected with the current test suite and it could be a problem related to other modules, I've changed the test suite to raise actually an error here (bb184efc06bfcd726ebe63851498c23873ee39e0)

So, as a summary, we're leaking that there's indeed an URL here. I don't know where it's exactly the problem.

alecslupu commented:
@andreslucena Specs are failing.

andreslucena commented:
> So, as a summary, we're leaking that there's indeed an URL here. I don't know where it's exactly the problem.

I finally found where was the problem with this. It's because `decidim-templates` was defining the admin routes in its own module instead of on `decidim-admin` as the rest of the modules, so it wasn't respecting the contracts from the admin.

In the process I've added more specs about the access, as I wanted to see if this was happening in other modules/sections. 

I've also applied a refactor (Extract Method) on a moderator check that I found while reading code.

I could leave only the relevant things for fixing the original bug, but I think that all of the changes make the code more clear and with more specs. 

This is ready to be reviewed again @alecslupu

alecslupu commented on decidim-admin/app/controllers/concerns/decidim/admin/needs_admin_tos_accepted.rb:
```diff
@@ -37,6 +38,41 @@ def permitted_paths
+      end
+
+      def participatory_process_user_role?
+        return false unless defined?(Decidim::ParticipatoryProcessUserRole)
```
I am not big fan of this coupling. Maybe we could have a manifest of some type that we could use for this kind of checks?

```suggestion
        return false unless Decidim.module_installed?(:participatory_process)
```

alecslupu commented on decidim-admin/app/controllers/concerns/decidim/admin/needs_admin_tos_accepted.rb:
```diff
@@ -37,6 +38,41 @@ def permitted_paths
+      end
+
+      def assembly_user_role?
+        return false unless defined?(Decidim::AssemblyUserRole)
```
```suggestion
        return false unless  Decidim.module_installed?(:assemblies)
```

alecslupu commented on decidim-admin/app/controllers/concerns/decidim/admin/needs_admin_tos_accepted.rb:
```diff
@@ -37,6 +38,41 @@ def permitted_paths
+      end
+
+      def conference_user_role?
+        return false unless defined?(Decidim::ConferenceUserRole)
```
```suggestion
        return false unless  Decidim.module_installed?(:conferences)
```

alecslupu commented on decidim-admin/app/controllers/concerns/decidim/admin/needs_admin_tos_accepted.rb:
```diff
@@ -37,6 +38,41 @@ def permitted_paths
+      end
+
+      def voting_monitoring_commitee_member?
+        return false unless defined?(Decidim::Votings::MonitoringCommitteeMember)
```
```suggestion
        return false unless  Decidim.module_installed?(:elections)
```

alecslupu commented on decidim-core/app/models/decidim/user.rb:
```diff
@@ -205,7 +205,7 @@ def tos_accepted?
 
     def admin_terms_accepted?
-      return true if admin_terms_accepted_at
+      true if admin_terms_accepted_at
```
```suggestion
      admin_terms_accepted_at.present?
```

alecslupu commented on decidim-admin/config/routes.rb:
```diff
@@ -39,6 +39,8 @@
       mount manifest.admin_engine, at: "/#{manifest.name}", as: "decidim_admin_#{manifest.name}"
     end
 
+    mount Decidim::Templates::AdminEngine, at: "/templates", as: "decidim_admin_templates" if defined?(Decidim::Templates)
```
Why this change? We should really avoid coupling modules in this manner.

alecslupu commented on decidim-templates/lib/decidim/templates/admin_engine.rb:
```diff
@@ -57,12 +57,6 @@ class AdminEngine < ::Rails::Engine
 
-      initializer "decidim_templates_admin.mount_routes" do
-        Decidim::Core::Engine.routes do
-          mount Decidim::Templates::AdminEngine, at: "/admin/templates", as: "decidim_admin_templates"
```
Why don't we use something like: 
```

      initializer "decidim_templates_admin.mount_routes" do
        Decidim::Admin::Engine.routes do
          mount Decidim::Templates::AdminEngine, at: "/templates", as: "decidim_admin_templates"
        end
      end
```

andreslucena replied:
I agree that introducing coupling here is a step back. 

Sadly, what you're proposing doesn't work, as it gives the behaviour that started this whole PR.

```ruby
$ bin/rspec decidim-admin/spec/system/admin_access_spec.rb

  1) AdminAccess when the user is a normal user behaves like a 404 page visiting the page leads to a 404
     Failure/Error: DEFAULT_FAILURE_NOTIFIER = lambda { |failure, _opts| raise failure }
       expected "/admin/" to equal "/admin/templates/questionnaire_templates"
     Shared Example Group: "a 404 page" called from ./spec/system/admin_access_spec.rb:29
     # /home/apereira/Work/decidim/decidim/decidim-core/lib/decidim/core/test/shared_examples/errors.rb:22:in `block (3 levels) in <top (required)>'

(...)

Finished in 5.9 seconds (files took 3.41 seconds to load)
7 examples, 1 failure

Failed examples:

rspec ./spec/system/admin_access_spec.rb:29 # AdminAccess when the user is a normal user behaves like a 404 page visiting the page leads to a 404

(...)
``` 

For respecting the `decidim-admin` module behaviour, we need to mount in their routes. Other modules don't have this problem because they're being mount by using a Registry/Manifest. For instance:

Spaces:

https://github.com/decidim/decidim/blob/a44de3d3c6bb6a6fd58dae7eae2a70f4b2ca731a/decidim-admin/config/routes.rb#L18-L20

Authorizations:

https://github.com/decidim/decidim/blob/a44de3d3c6bb6a6fd58dae7eae2a70f4b2ca731a/decidim-admin/config/routes.rb#L38-L40

As Templates isn't neither a Space nor an Authorization, then it doesn't implement any manifest that we can mount. 

What we could/should do if we don't want to introduce this coupling is to create a new Registry Manifest for a more generic kind of section in the Admin panel. But doing that would probably mean that this can't be a `type: fix` any more, so that's out of the scope of the current PR (IMO).

alecslupu replied:
Ok, let's mount it as it is, as we already have the dependency introduced by #10176
https://github.com/decidim/decidim/blob/61c0ff710b400fb7fed5242716077f3d63320112/decidim-admin/app/controllers/decidim/admin/application_controller.rb#L35-L37

alecslupu commented on decidim-admin/lib/decidim/admin/test/admin_participatory_space_access_examples.rb:
```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+shared_examples "showing the page" do
```
This example is too generic. "showing the page" is an example that could refer to : 
- Page component
- Static pages 
- Any particular page. 
- On a later stage, we may think this example as something trivial that answers to "is the current page loaded?"

andreslucena replied:
Yes, it's a lousy name. I've renamed it to ['accessing the participatory space'](https://github.com/decidim/decidim/pull/10968/commits/829f3a7138512129ab86198e7459e5c20255c365)

alecslupu reviewed (changes requested):
I just read the code. I think it would be a great addition, but i do have style, coupling related issues.

alecslupu commented:
@andreslucena can you fix the failing specs, please?

andreslucena commented:
The only failing spec is unrelated to this PR and fixed on #11587, so this is ready for another round

alecslupu commented:
@andreslucena 
I have checked this scenario ( last from your [comment](https://github.com/decidim/decidim/pull/10968#issuecomment-1647408497)): 

- Login as user@example.org
- Type in the URL /admin/assemblies/exaggerate-plain/edit
- See that you have a 404 page with this same URL
- Type in the URL /admin/templates/questionnaire_templates
- See that you have a 404 page - but now we're redirected to the /admin path.


But now, i get a `No route matches [GET] "/admin/templates/questionnaire_templates"`
![image](https://github.com/decidim/decidim/assets/105683/56ba2fa3-e96e-438e-a948-e952b09722f1)

From my point of view, considering that no user should be able to access admin related links, i would say this is ok.
